src/base/rss/rss_autodownloadrule.cpp

```diff
@@ -334,6 +372,20 @@ bool AutoDownloadRule::matches(const QString &articleTitle) const
         return false;
     }
 
+    if (useSmartFilter()) {
+        // now see if this episode has been downloaded before
+        QString episodeStr = computeEpisodeName(articleTitle);
+
+        if (!episodeStr.isEmpty()) {
+            bool previouslyMatched = m_dataPtr->previouslyMatchedEpisodes.contains(episodeStr);
+            bool isRepack = articleTitle.contains("REPACK", Qt::CaseInsensitive) || articleTitle.contains("PROPER", Qt::CaseInsensitive);
+            if (previouslyMatched && !isRepack)
+                return false;
+
+            m_dataPtr->lastComputedEpisode = episodeStr;
+        }
```

Comment:
IIRC, it allows to download many PROPER/REPACK copies of the same episode.

Reply:
I'm not quite sure what you mean. I suppose we could keep track of the repacks aswell, but you usually want to always download these.

Reply:
>I'm not quite sure what you mean.

Let's imagine the following articles sequence:
1. Some title s01e03
2. Some title s01e03
3. Some title s01e03 REPACK
4. Some title s01e03 REPACK
5. Some title s01e03 REPACK
It downloads 1st torrent, skips the 2nd (as expected), and download 3, 4 and 5th torrents. Is this behavior expected?

Reply:
I can pitch in and say yes, that is expected. The convention is that torrents labeled REPACK include some kind of fix that means it supersedes the previous torrent, and should always be downloaded.

Reply:
>it supersedes the previous torrent

Even other repacks?

Reply:
> Even other repacks?
Yes, I've just been testing with one where this has happened.

Reply:
>Yes, I've just been testing with one where this has happened.

I didn't understand what do you mean.
I was just asking is it normal (expected) that it downloads all Repack/Proper episode variants?

Reply:
>I was just asking is it normal (expected) that it downloads all Repack/Proper episode variants?

Yes, it is expected behaviour.

Reply:
>Yes, it is expected behaviour.

Why?

@elFarto, the main problem of this PR is your reluctance (or inability) to provide a detailed description (in some example) of the incident, processed by this feature, to not involved people could clearly understand the meaning of this process. Why duplicate episodes typically occur? Why, in contrast to the "normal" episodes, we don't have to skip repeated "REPACK/PROPER" episodes?
Now this feature is more like a "personal use" than the "public use".

Reply:
@elFarto, forgive my boredom in this issue, I just don't want to answer such questions of users (when we release this feature), not realizing why it works this way and not otherwise.

Reply:
@glassez I'll chime in again and try to help explain it in detail. Hopefully you'll find this satisfactory:

The labels "REPACK" and "PROPER" are common conventions used among release groups providing episodes of TV shows, as well as other media. These conventions are used among pretty much every bittorrent tracker that distributes these files. [Here is one link that goes into description of these various labels](https://www.dslreports.com/faq/10705), but in summary, "REPACK" means a rerelease of a previous episode, likely by the same group, meant to correct an error in the first copy. "PROPER" can mean the same thing, or more specifically a release by another group meant to correct an error from the first group.

These specific meanings may not be consistent among releases, but you always want to download torrents labeled PROPER or REPACK because they, quite consistently, only appear when the episode you've *already downloaded* contains an error. For example, the few times it's happened since I've been using this PR by @elFarto, the original file that was downloaded was broken to the point of being unplayable. But I didn't need to do anything or worry about it because the PROPER / REPACK version was also automatically downloaded. And as @elFarto has said, sometimes there's more than one because it takes the release groups a few tries to fix an error.

It's also worth mentioning that other torrent clients or RSS downloaders with a "smart episode" feature do this exact same thing. If you're concerned about users nagging you for support requests, then this will save you many questions from people asking for this exact feature.

Reply:
Well, I got the following picture:
1. There are several releasers (release groups) on some tracker.
2. They release the same episodes independently so there are duplicates can appear (otherwise this issue can't exist).
3. There is no acceptable (or reliable) way to separate the episodes of the various relisers (using download rules), or we don't want (unable) to do so for some other reason.
4. When some error is detected in the release the releaser publish fixed release (marked as REPACK or PROPER). There also can be some PROPER/REPACK release fixing error in previous PROPER/REPACK release (isn't it?).
5. Since we can't distinguish releases from different releasers (because of the same reasons as in Item 3) we have only to download them all.

What I am mistaken?

Reply:
@glassez That's mostly correct, but the whole picture is a little different than that.

1. Release groups are usually not associated with any particular tracker. It's the tracker's users that obtain the release and make it available on the tracker.
2. This is correct. Additionally, sometimes the users of a tracker make their own rip of an episode and make it available, alongside the scene releases.
3. This completely depends on the tracker in question. Some of them are very, very tightly organized and you can easily distinguish different releases from the RSS post's title. Other times this is not the case at all. In the end it is still the responsibility of the qBittorrent user to create a good rule definition in the RSS downloader so that it only considers torrents they want, even when using the smart episode filter.
4. This is correct. Additionally, a tracker's users can also create their own fixed version of a torrent. Either way, the torrent's title should have PROPER or REPACK in its title, as per convention.
5. Again, this depends on the tracker and how well organized it is, but generally the best way to make sure a good version of an episode is downloaded is first use the user's rule definition in the RSS downloader to figure out which episodes are candidates for downloading, then extract out the season/episode number from the RSS post and ensure it only downloads one copy. The exception of course is when PROPER or REPACK appears. It is technically possible that too many copies of the episode could be downloaded, but in practice this almost never happens and the feature works correctly.

I've been using this smart episode feature for over a year now, granted the version of the PR I merged in and compiled doesn't include the option to change or add regex's. However it has been working flawlessly, including when PROPER or REPACK torrents are downloaded.

Reply:
I'm going to chime in and say that - as a user, this is exactly what I'd want & expect. I came from uTorrent and that's how their smart episode filter works (granted they don't have the options for custom regex).